1. Basics/task_7/part_3/task_3_on_apache_handler.php

```diff
@@ -0,0 +1,172 @@
+	$correctNums = preg_match('#^(\d+,?)+$#', $nums);							// 400
+	
+	if(!$methodGet) return ['code' => 400, 'info' => 'Request method is invalid. The method GET is needed.'];
+
+	if(!$correctUri) return ['code' => 404, 'info' => 'Not found. URI is invalid.'];
+
+	if(!$nums) return ['code' => 400, 'info' => 'Value \'nums\' is absent.'];
+	
+	if(!$correctNums) return [ 'code' => 400, 'info' => 'Value \'nums\' is NOT comma separate numbers'];
+	
+	return false; // 200 OK	
+}
+
+// HTML OUTPUT FUNCTIONS //////////////////////////////////////////////////////////////////////////
+ /**
+  * @param string $responseBody
+  * 
+  * @return void
+  */
+ function outputHTML(string $responseBody) : void {
+
+	// send response for right headers list view on the HTML page 
+	flush();
+	
+	// title
+	echo '<h1>Task 3:</h1>';
+	
+	// request
+	echo '<h2>Request:</h2>';
+	outputRequest();	
+	echo '<hr style="border-top: 5px double #777">';
+	
+	// response
+	echo '<h2>Response:</h2>';
+	outputResponse($responseBody);
+	echo '<hr style="border-top: 5px double #777">';
+
+	// link 
+	echo '<br>'. mb_chr(129092) .' <a href=' . $_SERVER['HTTP_REFERER'] .'> Back to the request page</a>';
+}
+
+/**
+ * Output the user`s request uncluding some decorative elements
+ * 
+ * @return void
+ */
+function outputRequest() : void {
+	
+	// request line
+	echo testInputValue(apache_lookup_uri($_SERVER['REQUEST_URI'])->the_request) . '<br>';
+
+	// headers
+	$headers = getallheaders();
+	array_walk($headers, 'testInputValueByReference');
+	array_walk($headers, 'printKeyAndValue');
+
+	// body
+	$body = testInputValue(file_get_contents('php://input'));
+	if($body) echo '<br>' . $body;
+}
+
+/**
+ * @param bool|string $body
+ * 
+ * @return void
+ */
+function outputResponse(bool|string $body) : void {
+	$code = http_response_code();
+	echo $_SERVER['SERVER_PROTOCOL'] . ' ' . $code . ' ' . getResponseStatusMessage($code) . '<br>';
+	
+	$headers = apache_response_headers();
+	array_walk($headers, 'printKeyAndValue');
+	echo '<br>';
+	
+	if($body) echo $body;
+}
+
+/**
+ * @param int $statusCode
+ * 
+ * @return string
+ */
+function getResponseStatusMessage(int $statusCode) : string {
+    return RESPONSE_DATA[$statusCode]['statusMessage'];
+}
+
+/**
+ * Outputs key and its value in a nice separate line 
+ * (Uses as callback function for array_walk() for print header's/variable's names and values)
+ * 
+ * @param string $value
+ * @param string $key
+ * 
+ * @return void
+ */
+function printKeyAndValue(string &$value, string $key) : void {
+	echo '<strong>' . $key . '</strong>: ' . $value . '<br>';
+}
+
+// CHECK INPUT DATA FUNCTIONS //////////////////////////////////////////////////
+/**
+ * Changes input data to prevent harm, injections ect.
+ * (Able to use as a callback function for array_walk() and for array_walk_recursive())
+ * 
+ * @param string $data
+ * 
+ * @return void
+ */
+ function testInputValueByReference(string &$data) : void { //
+	$data = trim($data);
```

Comment:
and where is test?

Reply:
OK. `makeInputDataSafe...()` sounds better.